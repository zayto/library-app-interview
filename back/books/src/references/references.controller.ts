import { Controller } from '@nestjs/common';

@Controller('references')
export class ReferencesController {}
