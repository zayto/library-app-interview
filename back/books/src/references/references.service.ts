import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import {
  BookRefStatusEnum,
  IBookRefRequest,
  IBookRefResponse,
} from '../models/interfaces';
import { Reference, ReferenceDocument } from '../models/reference.schema';

@Injectable()
export class ReferencesService {
  constructor(
    @InjectModel(Reference.name) private refModel: Model<ReferenceDocument>,
  ) {}

  public async getBookReferenceById(
    refId: string,
  ): Promise<IBookRefResponse | null> {
    Logger.log(`GetBookReferenceById called with id=${refId}`);
    return await this.refModel.findById<IBookRefResponse>(refId);
  }

  public async getAllBooksRefs(): Promise<IBookRefResponse[]> {
    Logger.log('GetAllBookRefs called');
    return this.refModel.find().exec();
  }

  public async createBookRef(
    bookRef: IBookRefRequest,
  ): Promise<IBookRefResponse> {
    Logger.log(`CreateBookRef called with payload`, {
      bookRef: JSON.stringify(bookRef),
    });

    const createdBookRef = await this.refModel.create({
      ...bookRef,
      available: 0,
      totalQuantity: 0,
      status: BookRefStatusEnum.TO_ORDER_BACK, // Represents a new entry but with 0 quantity, we should order some books
    });
    return createdBookRef;
  }

  public async borrowBookAndUpdateBookRef(
    bookRef: IBookRefResponse,
  ): Promise<void> {
    const { _id: bookRefId } = bookRef;
    if (!this.isBookRefAvailable(bookRef)) {
      throw new Error(
        `Book with ref=${bookRefId} is not available to be borrowed`,
      );
    }

    const refToUpdate: Partial<IBookRefResponse> = {
      title: bookRef.title,
      author: bookRef.author,
      excerpt: bookRef.excerpt,
      status:
        bookRef.available > 0
          ? BookRefStatusEnum.AVAILABLE
          : BookRefStatusEnum.UNAVAILABLE,
      available: bookRef.available - 1,
    };
    // Update availability status and quantity after borrowing
    await this.refModel.findOneAndUpdate(bookRefId, refToUpdate);
  }

  public async addNewBookQuantity(
    bookRefId: string,
    quantity: number,
  ): Promise<IBookRefResponse> {
    if (quantity <= 0) {
      throw new Error(
        `Invalid quantity provided (quantity=${quantity}) to add for bookRef with id=${bookRefId}`,
      );
    }
    const bookRef = await this.refModel.findById(bookRefId, null);
    bookRef.available += quantity;
    bookRef.totalQuantity += quantity;
    if (bookRef.available > 0) {
      bookRef.status = BookRefStatusEnum.AVAILABLE;
    }
    const updated = await bookRef.save();
    Logger.log(
      `Added quantity ${quantity} to ref with id=${bookRefId}, new availableQuantity=${updated.available}, new totalQuantity=${updated.totalQuantity}`,
    );

    return updated;
  }

  private isBookRefAvailable(bookRef: IBookRefResponse): boolean {
    return (
      bookRef.available > 0 && BookRefStatusEnum.AVAILABLE === bookRef.status
    );
  }
}
