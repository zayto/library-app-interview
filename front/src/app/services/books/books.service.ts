import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { IBook } from 'src/app/types/interfaces';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root',
})
export class BooksService {
  constructor(private http: HttpClient) {}

  getBookRefs(): Observable<IBook[]> {
    const url = `${environment.apiUrl}/refs`;

    return this.http.get<IBook[]>(url);
  }
}
