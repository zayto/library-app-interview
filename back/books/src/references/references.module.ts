import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { ReferencesController } from './references.controller';
import { ReferencesService } from './references.service';
import { BookRef, BookRefSchema } from '../models/book-ref.schema';
import { ConfigModule } from '@nestjs/config';

@Module({
  imports: [
    ConfigModule.forRoot(),
    MongooseModule.forRoot(
      `mongodb://user:password@${process.env.MONGO_HOST}/library`,
    ),
    MongooseModule.forFeature([{ name: BookRef.name, schema: BookRefSchema }]),
  ],
  controllers: [ReferencesController],
  providers: [ReferencesService],
  exports: [MongooseModule],
})
export class ReferencesModule {}
