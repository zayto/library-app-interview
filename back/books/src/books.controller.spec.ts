import { BullModule } from '@nestjs/bull';
import { ClientsModule, Transport } from '@nestjs/microservices';
import { getModelToken } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';
import { Model, Types } from 'mongoose';
import { BooksController } from './books.controller';
import { BooksService } from './books.service';
import { Book } from './models/book.schema';
import {
  BookRefStatusEnum,
  BookStatusEnum,
  IBookRefResponse,
  IBookResponse,
} from './models/interfaces';
import { Reference } from './models/reference.schema';
import { ReferencesService } from './references/references.service';

describe('BooksController', () => {
  let booksController: BooksController;
  let bookModel: Model<Book>;
  let refModel: Model<Reference>;

  const mockBook: IBookResponse = {
    _id: new Types.ObjectId('6209880b6f43a7e034c84941'),
    reference: new Types.ObjectId('6209880b6f43a7e034c84943'),
    status: BookStatusEnum.AVAILABLE,
    owner: null,
  };

  const mockRef: IBookRefResponse = {
    _id: new Types.ObjectId('6209880b6f43a7e034c84943'),
    author: 'Author',
    title: 'Title',
    excerpt: 'Excerpt',
    available: 1,
    totalQuantity: 1,
    status: BookRefStatusEnum.AVAILABLE,
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ClientsModule.register([
          { name: 'USERS_SERVICE', transport: Transport.REDIS },
        ]),
        BullModule.registerQueue({
          name: 'returnedBooks',
        }),
      ],
      controllers: [BooksController],
      providers: [
        BooksService,
        ReferencesService,
        {
          provide: getModelToken(Book.name),
          useValue: {
            new: jest.fn().mockResolvedValue(mockBook),
            constructor: jest.fn().mockResolvedValue(mockBook),
            find: jest.fn().mockReturnValue({ exec: () => [mockBook] }),
            create: jest.fn(),
            exec: jest.fn(),
          },
        },
        {
          provide: getModelToken(Reference.name),
          useValue: {
            new: jest.fn().mockResolvedValue(mockRef),
            constructor: jest.fn().mockResolvedValue(mockRef),
            find: jest.fn().mockReturnValue({ exec: () => [mockRef] }),
            create: jest.fn(),
            exec: jest.fn(),
          },
        },
      ],
    }).compile();

    bookModel = module.get<Model<Book>>(getModelToken('Book'));
    refModel = module.get<Model<Reference>>(getModelToken('Reference'));

    booksController = module.get<BooksController>(BooksController);
  });

  describe('root', () => {
    it('should return all books', async () => {
      // TODO: Fix this test, need to properly mock the mongoose find/exec calls
      const books = await booksController.getAllBooks();
      expect(books).toMatchSnapshot();
    });
  });
});
