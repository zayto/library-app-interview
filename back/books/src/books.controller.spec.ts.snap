// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksController root should return all books 1`] = `
Array [
  Object {
    "_id": "6209880b6f43a7e034c84941",
    "owner": null,
    "reference": "6209880b6f43a7e034c84943",
    "status": "AVAILABLE",
  },
]
`;
